File: DotNetGrc/Grc/Cst/Visitor/GVNode.cs
using System;
using System.Collections.Generic;

namespace Grc.Cst.Visitor
{
	public class GVNode
	{
		private GVNode parent;

		private List<GVNode> children;

		private int id;

		private string data;

		public GVNode(int id, string data) : this(id)
		{
			this.data = data;
		}

		public GVNode(int id)
		{
			this.id = id;

			this.children = new List<GVNode>();
		}

		public virtual void addChild(GVNode child)
		{
			child.parent = this;

			this.children.Add(child);
		}

		public virtual void print()
		{
			Console.WriteLine(this.id);
			Console.WriteLine(this.data);

			foreach (GVNode c in children)
				c.print();
		}

		public virtual void printRelations()
		{
			foreach (GVNode c in children)
			{
				Console.WriteLine(this.id + " -> " + c.id);

				c.printRelations();
			}
		}

		public virtual void printGraphViz()
		{
			// pre
			if (this.parent != null)
			{
				Console.WriteLine("\t" + this.gvName() + " ;");

				Console.WriteLine("\t" + this.gvName() + " [label=\"" + gvData() + "\"] ;");
			}

			// in
			foreach (GVNode c in children)
			{
				if (this.parent != null)
					Console.WriteLine("\t" + this.gvName() + " -- " + c.gvName() + " ;");

				c.printGraphViz();
			}

			// post
		}

		public virtual string gvName()
		{
			return string.Format("n{0:D4}", this.id);
		}

		public virtual string gvData()
		{
			return data.Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("[", "\\[").Replace("]", "\\]");
		}
	}
}

File: Parser/src/main/java/k31/grc/cst/visitor/GVNode.java
package k31.grc.cst.visitor;

import java.util.LinkedList;
import java.util.List;

public class GVNode {

	@SuppressWarnings("unused")
	private GVNode parent;

	private List<GVNode> children;

	@SuppressWarnings("unused")
	private int id;

	private String data;

	public GVNode(int id, String data) {

		this(id);

		this.data = data;
	}

	public GVNode(int id) {

		this.id = id;

		this.children = new LinkedList<GVNode>();
	}

	public void addChild(GVNode child) {

		child.parent = this;
		this.children.add(child);
	}

	public void print() {

		System.out.println(this.id);
		System.out.println(this.data);

		for (GVNode c : children) {
			c.print();
		}
	}

	public void printRelations() {

		for (GVNode c : children) {

			System.out.println(this.id + " -> " + c.id);

			c.printRelations();
		}

	}

	public void printGraphViz() {

		// pre
		if (this.parent != null) {

			System.out.println("\t" + this.gvName() + " ;");

			System.out.println("\t" + this.gvName() + " [label=\"" + gvData() + "\"] ;");
		}

		// in
		for (GVNode c : children) {

			if (this.parent != null)
				System.out.println("\t" + this.gvName() + " -- " + c.gvName() + " ;");

			c.printGraphViz();
		}

		// post

	}

	public String gvName() {

		return String.format("n%1$04d", this.id);
	}

	public String gvData() {

		return data
				.replace("\\", "\\\\")
				.replace("\"", "\\\"")
				.replace("[", "\\[")
				.replace("]", "\\]");
	}
}
